pkg/kubelet/kubelet.go

```diff
@@ -1152,120 +1152,183 @@ func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Co
-		if err != nil {
-			return err
-		}
-		podInfraContainerID, err = kl.createPodInfraContainer(pod)
-		if err != nil {
-			glog.Errorf("Failed to introspect pod infra container: %v; Skipping pod %q", err, podFullName)
-			return err
-		}
-		if count > 0 {
-			// Re-list everything, otherwise we'll think we're ok.
-			containersInPod, err = dockertools.GetKubeletDockerContainers(kl.dockerClient, false)
-			if err != nil {
-				glog.Errorf("Error listing containers %#v", containersInPod)
-				return err
-			}
-		}
+	if found {
+		glog.V(4).Infof("Found infra pod for %q", podFullName)
+		containersToKeep[podInfraContainerID] = 0
 		podStatus, err = kl.GetPodStatus(podFullName, uid)
 		if err != nil {
 			glog.Errorf("Unable to get pod with name %q and uid %q info with error(%v)", podFullName, uid, err)
 		}
 	} else {
-		podStatus, err = kl.GetPodStatus(podFullName, uid)
-		if err != nil {
-			glog.Errorf("Unable to get pod with name %q and uid %q info with error(%v)", podFullName, uid, err)
-		}
-	}
-	containersInPod.RemoveContainerWithID(podInfraContainerID)
-
-	ref, err := api.GetReference(pod)
-	if err != nil {
-		glog.Errorf("Couldn't make a ref to pod %q: '%v'", podFullName, err)
-	}
-
-	podVolumes, err := kl.mountExternalVolumes(pod)
-	if err != nil {
-		if ref != nil {
-			kl.recorder.Eventf(ref, "failedMount",
-				"Unable to mount volumes for pod %q: %v", podFullName, err)
-		}
-		glog.Errorf("Unable to mount volumes for pod %q: %v; skipping pod", podFullName, err)
-		return err
+		glog.V(2).Infof("No Infra Container for %q found. All containers will be restarted.", podFullName)
+		createPodInfraContainer = true
 	}
 
-	for _, container := range pod.Spec.Containers {
+	for index, container := range pod.Spec.Containers {
 		expectedHash := dockertools.HashContainer(&container)
 		dockerContainerName := dockertools.BuildDockerName(uid, podFullName, &container)
+		// If we're creating infra containere everything will be killed anyway
 		if dockerContainer, found, hash := containersInPod.FindPodContainer(podFullName, uid, container.Name); found {
 			containerID := dockertools.DockerID(dockerContainer.ID)
 			glog.V(3).Infof("pod %q container %q exists as %v", podFullName, container.Name, containerID)
 
-			// look for changes in the container.
-			containerChanged := hash != 0 && hash != expectedHash
-			if !containerChanged {
-				result, err := kl.probeContainer(pod, podStatus, container, dockerContainer)
-				if err != nil {
-					containersInPod.RemoveContainerWithID(containerID)
-					continue
-				}
-				if result == probe.Success {
-					containersInPod.RemoveContainerWithID(containerID)
-					continue
+			if !createPodInfraContainer {
+				// look for changes in the container.
+				containerChanged := hash != 0 && hash != expectedHash
+				if !containerChanged {
+					result, err := kl.probeContainer(pod, podStatus, container, dockerContainer)
+					if err != nil {
+						// I don't get this logic...
```

Comment:
Yes, but I think that it should strongly depend on error type. Some of them may be transient, but I guess some of them are not.